node-metrics/src/api/node_validator/v0/mod.rs

```diff
@@ -540,28 +553,93 @@ impl ProcessProduceLeafStreamTask {
+    where
+        R: LeafStreamRetriever<Item = Leaf<SeqTypes>>,
+        K: Sink<Leaf<SeqTypes>, Error = SendError> + Clone + Send + Sync + Unpin + 'static,
+    {
+        // We want to try and ensure that we are connected to the HotShot Query
+        // Service, and are consuming leaves.
+        // - If we are able to connect, then we can start relaying Leaves into
+        //   the leaf sender.  If we are unable
+        // - If we are not able to connect, then we should sleep, and retry
+        //   until we are able to reconnect.  Each failure **should** make some
+        //   noise so that if we are never able to reconnect, we are at least
+        //   telling someone about it.
+        // - If the task for consuming the leaves completes, then we should
+        //   also attempt to reestablish the connection to start consuming
+        //   the leave again.
+
+        loop {
+            // Retrieve a stream
+            let Ok(stream) = Self::retrieve_leaf_stream(&leaf_stream_retriever).await else {
+                panic!("failed to retrieve leaf stream");
+            };
+
+            // Consume the leaves of a stream
+            Self::process_consume_leaf_stream::<R, K>(stream, leaf_sender.clone()).await;
+            tracing::warn!("leaf stream ended, will attempt to re-acquire leaf stream");
+        }
+    }
+
+    /// [retrieve_leaf_stream] attempts to retrieve the Stream of Leaves from
+    /// the given [LeafStreamRetriever].
+    ///
+    /// This function will loop on failure until it is able to retrieve the
+    /// [Stream].  This does mean that it could potentially get in a state
+    /// where it can loop indefinitely.
+    ///
+    /// This function also implements exponential backoff with a maximum
+    /// delay of 5 seconds.
+    async fn retrieve_leaf_stream<R>(
+        leaf_stream_receiver: &R,
+    ) -> Result<R::Stream, RetrieveLeafStreamError>
+    where
+        R: LeafStreamRetriever<Item = Leaf<SeqTypes>>,
+    {
+        for attempt in 1..=100 {
+            let leaves_stream_result = leaf_stream_receiver.retrieve_stream(None).await;
+
+            let leaves_stream = match leaves_stream_result {
+                Err(error) => {
+                    // We failed to retrieve the stream. We will try again, but we
+                    // should sleep for a bit before so as not to overwhelm the
+                    // service.
+                    tracing::warn!(
+                        "attempt {attempt} to connect to leaf stream failed with error {error}"
+                    );
+
+                    // Our retry penalty will be a minimum of 100ms, and a maximum
+                    // of 5 seconds.
+                    // For every failed iteration, we will double our delay, up
+                    // to the maximum of 5 seconds.
+
+                    let retry_penalty = max(
+                        Duration::from_millis(100) * 2u32.pow(attempt),
```

Comment:
if attempt is large like 99 then wouldn't this result in an overflow?

Reply:
Also we have `BackoffParams` so maybe we can re use it here

Reply:
Hmm... the `BackoffParams` seem to indicate that they are used specifically for catchup.  At least that's what the environment variables indicate.  We could still potentially use the default here though.

Also, yeah, you're right... it would overflow as soon as it hits 30 attempts or so.

Reply:
`BackOffParams`'s api seems inconvenient at the moment as well.  It assumes an underlying that is being mutated, and requires `anyhow` to operate.

Reply:
Though it does look like we might be able to achieve what `BackOffParams::retry` is doing with `backoff` itself.

Reply:
ok lets leave BackOffParams